internal/cmd/manager/instance/upgrade/prepare/cmd.go

```diff
@@ -0,0 +1,133 @@
+func NewCmd() *cobra.Command {
+	var pgConfig string
+
+	cmd := cobra.Command{
+		Use:  "prepare [target]",
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			contextLogger := log.FromContext(cmd.Context())
+			dest := args[0]
+
+			if err := copyPostgresInstallation(cmd.Context(), pgConfig, dest); err != nil {
+				contextLogger.Error(err, "Failed to copy the PostgreSQL installation")
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&pgConfig, "pg-config", getEnvOrDefault("PG_CONFIG", "pg_config"),
+		`The path of "pg_config" executable. Defaults to "pg_config".`)
+
+	return &cmd
+}
+
+func getEnvOrDefault(env, def string) string {
+	if value, ok := os.LookupEnv(env); ok {
+		return value
+	}
+	return def
+}
+
+// copyPostgresInstallation replicates the PostgreSQL installation to the specified destination directory
+// for use by the ph_upgrade command as the old binary directory.
+//
+// Steps performed:
+// 1. Removes the existing destination directory if it exists.
+// 2. Retrieves the PostgreSQL binary, library, and shared directories using pg_config.
+// 3. Creates the corresponding directories in the destination path.
+// 4. Copies the contents of the PostgreSQL directories to the destination.
+// 5. Creates a bindir.txt file in the destination directory with the path to the binary directory.
+func copyPostgresInstallation(ctx context.Context, pgConfig string, dest string) error {
+	contextLogger := log.FromContext(ctx)
+
+	dest = path.Clean(dest)
+
+	contextLogger.Info("Copying the PostgreSQL installation to the destination", "destination", dest)
+
+	contextLogger.Info("Removing the destination directory", "directory", dest)
+	err := os.RemoveAll(dest)
+	if err != nil {
+		return fmt.Errorf("failed to remove the directory: %w", err)
+	}
+
+	contextLogger.Info("Creating the destination directory", "directory", dest)
+	err = os.MkdirAll(dest, 0o750)
+	if err != nil {
+		return fmt.Errorf("failed to create the directory: %w", err)
+	}
+
+	for _, config := range []string{"bindir", "pkglibdir", "sharedir"} {
+		sourceDir, err := getPostgresConfig(pgConfig, config)
+		if err != nil {
+			return err
+		}
+		sourceDir = path.Clean(sourceDir)
+		destDir := path.Clean(path.Join(dest, sourceDir))
+
+		if config == "bindir" {
+			destFile := path.Join(dest, "bindir.txt")
+			contextLogger.Info("Creating the bindir.txt file", "file", destFile)
+			_, err := fileutils.WriteStringToFile(destFile, fmt.Sprintf("%s\n", destDir))
+			if err != nil {
+				return fmt.Errorf("failed to write the %q file: %w", destFile, err)
+			}
+		}
+
+		contextLogger.Info("Creating the directory", "directory", destDir)
+		err = os.MkdirAll(destDir, 0o750)
+		if err != nil {
+			return fmt.Errorf("failed to create the directory: %w", err)
+		}
+
+		contextLogger.Info("Copying the files", "source", sourceDir, "destination", destDir)
+		err = exec.Command("cp", "-a", sourceDir+"/.", destDir).Run() //nolint:gosec
```

Comment:
This could possibly be sped up with the `--reflink=auto` flag depending on the filesystem used in the pvc. Reflinking whenever possible to make copies instant is the default behaviour of cp in coreutils 9.x, but not on 8.x which is still widely used.